src/main/ApplicationImpl.cpp

```diff
@@ -157,61 +157,98 @@ maybeRebuildLedger(Application& app, bool applyBuckets)
+            LOG_INFO(DEFAULT_LOG, "Dropping liquiditypools");
+            app.getLedgerTxnRoot().dropLiquidityPools(/*rebuild=*/false);
 #ifdef ENABLE_NEXT_PROTOCOL_VERSION_UNSAFE_FOR_PRODUCTION
-            case CONTRACT_DATA:
-                LOG_INFO(DEFAULT_LOG, "Dropping contractdata");
-                app.getLedgerTxnRoot().dropContractData();
-                break;
-            case CONFIG_SETTING:
-                LOG_INFO(DEFAULT_LOG, "Dropping configsettings");
-                app.getLedgerTxnRoot().dropConfigSettings();
-                break;
+            LOG_INFO(DEFAULT_LOG, "Dropping contractdata");
+            app.getLedgerTxnRoot().dropContractData(/*rebuild=*/false);
+            LOG_INFO(DEFAULT_LOG, "Dropping configsettings");
+            app.getLedgerTxnRoot().dropConfigSettings(/*rebuild=*/false);
 #endif
-            default:
-                abort();
+        }
+        else
+        {
+            for (auto let : toRebuild)
+            {
+                switch (let)
+                {
+                case ACCOUNT:
+                    LOG_INFO(DEFAULT_LOG, "Dropping accounts");
+                    app.getLedgerTxnRoot().dropAccounts(/*rebuild=*/true);
+                    break;
+                case TRUSTLINE:
+                    LOG_INFO(DEFAULT_LOG, "Dropping trustlines");
+                    app.getLedgerTxnRoot().dropTrustLines(/*rebuild=*/true);
+                    break;
+                case OFFER:
+                    LOG_INFO(DEFAULT_LOG, "Dropping offers");
+                    app.getLedgerTxnRoot().dropOffers(/*rebuild=*/true);
+                    break;
+                case DATA:
+                    LOG_INFO(DEFAULT_LOG, "Dropping accountdata");
+                    app.getLedgerTxnRoot().dropData(/*rebuild=*/true);
+                    break;
+                case CLAIMABLE_BALANCE:
+                    LOG_INFO(DEFAULT_LOG, "Dropping claimablebalances");
+                    app.getLedgerTxnRoot().dropClaimableBalances(
+                        /*rebuild=*/true);
+                    break;
+                case LIQUIDITY_POOL:
+                    LOG_INFO(DEFAULT_LOG, "Dropping liquiditypools");
+                    app.getLedgerTxnRoot().dropLiquidityPools(/*rebuild=*/true);
+                    break;
+#ifdef ENABLE_NEXT_PROTOCOL_VERSION_UNSAFE_FOR_PRODUCTION
+                case CONTRACT_DATA:
+                    LOG_INFO(DEFAULT_LOG, "Dropping contractdata");
+                    app.getLedgerTxnRoot().dropContractData(/*rebuild=*/true);
+                    break;
+                case CONFIG_SETTING:
+                    LOG_INFO(DEFAULT_LOG, "Dropping configsettings");
+                    app.getLedgerTxnRoot().dropConfigSettings(/*rebuild=*/true);
+                    break;
+#endif
+                default:
+                    abort();
+                }
             }
         }
 
-        tx.commit();
+        if (toRebuild.empty())
```

Comment:
I'm trying to preserve the original logic flow of the function here. Initially, this `toRebuild.empty()` check was at the beginning of the function and returned early. We don't want to do that now since we want to always drop the SQL tables if bucketlist loads are enabled, but I still want to exit early in that case so we don't hit the `applyBuckets` codepath.

Reply:
sure but there's no need to duplicate the commit call though - you can commit the tx regardless and exit early if there's nothing to rebuild (also, related to my other comment - if you setup toRebuild such that it takes the experimental flag into consideration , then there's no need to check the flag in the rest of the function, so you could use the same condition as before `toRebuild.empty()`)